File: javascript/core/src/verbs/pivot.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { PivotArgs } from '@datashaper/schema'

import type { ColumnTableStep } from './util/factories.js'
import { stepVerbFactory } from './util/factories.js'
import { singleExpression } from './util/index.js'

export const pivotStep: ColumnTableStep<PivotArgs> = (
	input,
	{ key, value, operation },
) => input.pivot(key, { [value]: singleExpression(value, operation) })

export const pivot = stepVerbFactory(pivotStep)

File: python/datashaper/datashaper/engine/verbs/pivot.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Pivot verb implementation."""

from typing import Any, cast

import pandas as pd

from datashaper.engine.create_verb_result import create_verb_result
from datashaper.engine.types import VerbResult
from datashaper.engine.verbs.verb_input import VerbInput
from datashaper.engine.verbs.verbs_mapping import verb
from datashaper.verbs import pivot


@verb(name="pivot", treats_input_tables_as_immutable=True)
def pivot_verb(
    input: VerbInput,
    **kwargs: Any,
) -> VerbResult:
    """Pivot verb implementation."""
    result = pivot(cast(pd.DataFrame, input.get_input()), **kwargs)
    return create_verb_result(result)
